admin/tabs/roles/RolesModel.ts

```diff
@@ -0,0 +1,280 @@
+            if (loadSpec.isStale) return;
+            this.setRoles(this.processRolesFromServer(data));
+        } catch (e) {
+            if (loadSpec.isStale) return;
+            XH.handleException(e);
+            this.clear();
+        }
+    }
+
+    selectRole(name: string) {
+        this.gridModel.selectAsync(name);
+    }
+
+    @action
+    setRoles(roles: HoistRole[]) {
+        this.allRoles = roles;
+        this.gridModel.loadData(roles);
+    }
+
+    @action
+    clear() {
+        this.allRoles = [];
+        this.gridModel.clear();
+    }
+
+    // -------------------------------
+    // Reactions
+    // -------------------------------
+
+    private groupByCategoryReaction(): ReactionSpec<boolean> {
+        const {gridModel} = this;
+        return {
+            track: () => this.groupByCategory,
+            run: groupByCategory => {
+                gridModel.setGroupBy(groupByCategory ? 'category' : null);
+                gridModel.setColumnVisible('category', !groupByCategory);
+                gridModel.autosizeAsync();
+            },
+            fireImmediately: true
+        };
+    }
+
+    async deleteAsync(role: HoistRole): Promise<boolean> {
+        const {effectiveUsers, effectiveDirectoryGroups, effectiveRoles} = role,
+            userCount = effectiveUsers.length,
+            groupCount = effectiveDirectoryGroups.length,
+            roleCount = effectiveRoles.length;
+
+        if (userCount || groupCount || roleCount) {
+            const confirm = await XH.confirm({
+                icon: Icon.warning(),
+                title: 'Confirm delete?',
+                message: div(
+                    'Deleting this role will affect:',
+                    ul(
+                        compact([
+                            userCount && pluralize('user', userCount, true),
+                            groupCount && pluralize('group', groupCount, true),
+                            roleCount && pluralize('role', roleCount, true)
+                        ]).map(it => li(it))
+                    ),
+                    'Are you sure you wish to continue?'
+                ),
+                confirmProps: {intent: 'danger', text: 'Confirm Delete'}
+            });
+            if (!confirm) return false;
+        }
+        await XH.fetchJson({
+            url: `roleAdmin/delete/${role.name}`,
+            method: 'DELETE'
+        });
+        this.refreshAsync();
+        return true;
+    }
+
+    // -------------------------------
+    // Implementation
+    // -------------------------------
+
+    private processRolesFromServer(
+        roles: Omit<HoistRole, 'users' | 'directoryGroups' | 'roles'>[]
+    ): HoistRole[] {
+        return roles.map(role => {
+            const membersByType = mapValues(groupBy(role.members, 'type'), members =>
+                members.map(member => member.name)
+            );
+            return {
+                ...role,
```

Comment:
Interesting general question of how concerned we are with getting the types exactly right - in this case note that nothing is converting the date fields, until they are loaded into store records, at which point the record data has a date. Could parse here for correctness, don't really think it matters more raising as a general discussion point....